daml-assistant/daml-helper/src/DamlHelper.hs

```diff
@@ -204,8 +209,13 @@ runStart = withProjectRoot $ \_ -> do
             writeFileUTF8 navigatorConfPath (T.unpack $ navigatorConfig parties)
             withNavigator sandboxPort navigatorPort navigatorConfPath [] $ \navigatorPh ->
                 void $ race (waitForProcess navigatorPh) (waitForProcess sandboxPh)
-    where sandboxPort = SandboxPort 6865
-          navigatorPort = NavigatorPort 7500
+
+-- there is a same function in port-util lib but does not expose it.
+getFreePort :: IO Int
+getFreePort = do
```

Comment:
This has a race condition: The port can be used by something else in between `openFreePort` giving you the port and sandbox/navigator listening on that port. In fact you could even end up with `sandboxP` and `navigatorP` pointing to the same port. At least Linux and MacOS generally don’t try to reuse ports aggressively so this usually works (and is in fact what we use in integration tests) but we don’t get any guarantees about it.

The proper solution here would be to ask sandbox and navigator to start on port `0` which will make them pick a free port automatically and then parse their output to figure out the actual port they are listening on. This is quite a bit more complex since you need to redirect the output of sandbox and navigator but imho we should do that anyway to make it clearer which output comes from sandbox and which output comes from navigator (e.g. by prefixing it with `[sandbox]` and `[navigator]`) and avoid interleaved output.

Reply:
Redirecting & tagging output as [sandbox] / [navigator] seems like a worthwhile thing to do, although a separate issue.

Solving the port problem / race condition properly requires additions to sandbox and navigator. I created the issue #833 to track that. In the mean time, I'm ok with the imprecise racey solution, but @neil-da is not too happy to change these multiple times, so I guess we should wait for the runtime team to solve #833.

Reply:
Yes, that looks like a proper fix, will wait till we have this fixed in sandbox and navigator.